plugins/amazonq/shared/jetbrains-community/src/software/aws/toolkits/jetbrains/services/amazonq/lsp/textdocument/TextDocumentServiceHandler.kt

```diff
@@ -0,0 +1,70 @@
+// Copyright 2025 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package software.aws.toolkits.jetbrains.services.amazonq.lsp.textdocument
+
+import com.intellij.openapi.Disposable
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.fileEditor.FileEditorManager
+import com.intellij.openapi.fileEditor.FileEditorManagerListener
+import com.intellij.openapi.vfs.VirtualFile
+import org.eclipse.lsp4j.DidCloseTextDocumentParams
+import org.eclipse.lsp4j.DidOpenTextDocumentParams
+import org.eclipse.lsp4j.TextDocumentIdentifier
+import org.eclipse.lsp4j.TextDocumentItem
+import software.aws.toolkits.jetbrains.services.amazonq.lsp.AmazonQLanguageServer
+
+class TextDocumentServiceHandler(
+    private val project: Project,
+    private val languageServer: AmazonQLanguageServer,
+    private val serverInstance: Disposable
```

Comment:
these two are linked so maybe it's cleaner to do

```kt
        fun executeIfRunning(project: Project, runnable: (AmazonQLanguageServer) -> Unit)  =
            getInstance(project).instance?.languageServer?.let { runnable(it) }
```
[...]
```kt
            override fun fileOpened(
                source: FileEditorManager,
                file: VirtualFile
            ) {
                executeIfRunning(file.project) {
                    it.textDocumentService.didOpen(...)
                }
            }
```

and have the listener registered globally through the extension point variant of this listener?